File: crates/cxx-qt-lib/src/core/qtlogging.cpp
// clang-format off
// SPDX-FileCopyrightText: 2025 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Joshua Goins <joshua.goins@kdab.com>
// SPDX-FileContributor: Joshua Booth <joshua.n.booth@gmail.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/qtlogging.h"

#include <QtCore/qlogging.h>

namespace rust {
namespace cxxqtlib1 {

inline void
log(QtMsgType type,
    const char* fileName,
    int lineNumber,
    const QString& message)
{
  qt_message_output(
    type,
    QMessageLogContext(fileName, lineNumber, nullptr, "default"),
    message);
}

void
q_debug(const char* fileName, int lineNumber, const QString& message)
{
  log(QtMsgType::QtDebugMsg, fileName, lineNumber, message);
}

void
q_info(const char* fileName, int lineNumber, const QString& message)
{
  log(QtMsgType::QtInfoMsg, fileName, lineNumber, message);
}

void
q_warning(const char* fileName, int lineNumber, const QString& message)
{
  log(QtMsgType::QtWarningMsg, fileName, lineNumber, message);
}

void
q_critical(const char* fileName, int lineNumber, const QString& message)
{
  log(QtMsgType::QtCriticalMsg, fileName, lineNumber, message);
}

void
q_fatal(const char* fileName, int lineNumber, const QString& message)
{
  log(QtMsgType::QtFatalMsg, fileName, lineNumber, message);
}

}
}

File: tests/qt_types_standalone/rust/src/qtlogging.rs
// SPDX-FileCopyrightText: 2025 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Joshua Booth <joshua.n.booth@gmail.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::{q_info, QString};

#[cxx::bridge]
mod ffi {
    extern "C++" {
        include!("cxx-qt-lib/qstring.h");
        type QString = cxx_qt_lib::QString;
    }

    extern "Rust" {
        fn log_info(message: &QString) -> QString;
    }
}

fn log_info(message: &QString) -> QString {
    q_info!("Message: {message}");
    QString::from(&format!("{}:{} - Message: {message}", file!(), line!() - 1))
}
